### DISCUSSION TITLE
How do I submit an empty array?

### DISCUSSION THREAD

**Discussion by timdavish** [react-hook-form/react-hook-form#3003] [Q&A] [answered]
[CodeSandbox](https://codesandbox.io/s/react-hook-form-usefieldarray-template-forked-tnqz6?file=/src/index.js)

1. Press "SUBMIT"
2. Console shows `data` as `test: [{ firstName: "Bill", lastName: "Luo" }]`
3. Press "DELETE"
4. Press "SUBMIT"
5. Console shows `data` as `{}`

Is there any natural way I can get `data` to be `{test: []}`?

Currently, I'm merging data with a default value, like this:
```ts
const defaultData = {
    test: [],
}

const fullData = {...defaultData, ...data}
```

But this does not scale well when I have field arrays inside of field arrays.

I've tried using `setValue` when `fields.length === 0`, but it appears you cannot set any form values to empty strings or empty objects.

**Comment by bluebill1049** [comment, marked as answer]
perhaps try useFieldArray

**Reply by timdavish** [reply]
In what way? I already am using it :)
The problem is the fact that once I delete the last item in the array, the entire array gets completely deleted from the form data.

**Reply by bluebill1049** [reply]
oh oops, sorry many questions, took a glance. I think we are doing a transform during submit, I guess you may have to do a data check on submitted data?

**Reply by timdavish** [reply]
Do you mean that I should check the `data` in the submit handler for the array, and if it isn't there, manually add it? Ff so, that's the solution I've already been going with. It would be very convenient to be able to at least configure a field array to allow empty arrays, though.

**Reply by bluebill1049** [reply]
yea, unfortunately, that's the thing react hook form's input values live in the inputs itself, so when all the inputs all get removed, it's will return empty data (just like the native form's behavior). probably have to do a `useEffect` to set an empty value for your field array. I mean we could do a fallback for an empty array, but it would lead to a breaking change to all the existing users.

**Reply by timdavish** [reply]
> probably have to do a useEffect to set an empty value for your field array

Can you elaborate a little bit on what you mean by this?

> I mean we could do a fallback for an empty array, but it would lead to a breaking change to all the existing users.

Yeah, this is fair. Putting this functionality behind a `useFieldArray` option or something similar could alleviate this worry though, right?

It may honestly make sense to keep the current behavior exactly how it currently is, especially if this isn't something a lot of users run into.. but I thought it was worth bringing up. In my case, I have nested `useFieldArray`s, and I need the data to not collapse/disappear out of the `data` object during submit, so while the `useEffect` solution does work, it gets a little awkward with my nested form data shape (it took me a bit to come up with a solution that I liked).

**Reply by bluebill1049** [reply]
I agree the current behavior may not be intuitive for developers who are used with controlled forms. As today, we don't have plans to have more options for useFieldArray just yet, if we going to introduce similar behavior, we probably doing the `array.length` check and return empty [],  which is similar to do that onSubmit function, will keep you updated if we going to introduce extra config tho. will leave a note on the project board as well.

**Reply by timdavish** [reply]
Sounds good. Thanks for considering this! And thanks for your work on this awesome package :)

**Reply by bluebill1049** [reply]
My pleasure.

**Reply by bluebill1049** [reply]
I have a good solution for this, by set `shouldUnregister` to `false`, we can keep an empty field array at the submit.

**Reply by daedmunoz** [reply]
I have the same issue as @timdavish. Even though setting `shouldUnregister` to `false` worked for me, somehow adding to and removing from arrays got really, really slow, and also a rich text editor (draft-js) component got broken. So, I kept `shouldUnregister` as `true` and validated the existence of the array field (I'm using Yup for validation).

**Reply by bluebill1049** [reply]
This is getting removed by V7. remove items will keep FieldArray as  `[]`

**Reply by prince272** [reply]
This solution does not work starting from react hook version 6.9.3 and above which is relating to the same problem.
https://github.com/react-hook-form/react-hook-form/discussions/1886

**Reply by prince272** [reply]
> This is getting removed by V7. remove items will keep FieldArray as `[]`

Will the update be apply to V6? Please do!

**Reply by bluebill1049** [reply]
@prince272 will try after v7 is released :)
